### ISSUE TITLE
[Typescript] `Socket` from `socket.io` cannot be unioned with `Socket` from `socket.io-client`

### ISSUE THREAD

**Issue by evanpurkhiser** [socketio/socket.io#3872]
**Describe the bug**

In 3.x of socket.io I could do the following:

```ts
import {Socket as ServerSocket} from 'socket.io';
import {Socket as ClientSocket} from 'socket.io-client';

function doStuff(ws: ClientSocket | ServerSocket) => {
	ws.on('some-event', () => {
		// ...
	});
}
```

In 4.x this now produces an error on the `ws.on` call:

```
[tsserver 2349] [E] This expression is not callable.
  Each member of the union type '(<Ev extends string>(ev: Ev, listener: ReservedOrUserListener<SocketReservedEvents, DefaultEventsMap, Ev>) => Socket<...>) | (<Ev extends string>(ev: Ev, listener: FallbackToUntypedListener<...>) => Socket<...>)' has signatures, but none of those signatures are compatible with each other.
```

**To Reproduce**

Socket.IO server / client version: `4.0.0`

See above

**Expected behavior**
I realize this was a major version bump between 3 and 4, so I can understand if this issue needs to be closed.

This also might be more of a typescript problem than anything else. I know the templated emitter feature was added in 4 (which I am looking forward to using once I get around this!) which I assume is what has made the method signatures incompatible?

I noticed I can "fix" this by doing

```tsx
const doStuff = (ws: ClientSocket | ServerSocket) => {
  if (ws instanceof ClientSocket) {
    ws.on('some-event', () => {
      // ...
    });
  }

  if (ws instanceof ServerSocket) {
    ws.on('some-event', () => {
      // ...
    });
  }
};
```

which is pretty goofy.

**Additional context**
I am on Typescript 4.0.2

**Comment by darrachequesne** [comment]
I could indeed reproduce the issue. Looks like it's related to https://github.com/socketio/socket.io/commit/0107510ba8a0f148c78029d8be8919b350feb633, but I'm not sure how we can fix this.

**Comment by darrachequesne** [comment]
For future readers:

It seems this was fixed in TypeScript between 4.1 and 4.2:

```
$ npm i typescript@4.1

changed 1 package, and audited 86 packages in 904ms

2 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

$ npx tsc
test.ts:5:8 - error TS2349: This expression is not callable.
  Each member of the union type '(<Ev extends string>(ev: Ev, listener: FallbackToUntypedListener<Ev extends "connect" | "connect_error" | "disconnect" ? SocketReservedEvents[Ev] : Ev extends string ? (...args: any[]) => void : never>) => Socket<...>) | (<Ev extends string>(ev: Ev, listener: FallbackToUntypedListener<...>) => Socket<...>)' has signatures, but none of those signatures are compatible with each other.

5     ws.on('some-event', () => {
         ~~


Found 1 error.

$ npm i typescript@4.2

changed 1 package, and audited 86 packages in 924ms

2 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

$ npx tsc
$
```

Please reopen if needed!
